### PR TITLE
Be a little friendlier to the Reddit API

### LINKED ISSUE #124: Reddit Integration: 429 Too Many Requests Errors
As of 09/17/2021, the Reddit integration has stopped working properly, as the API seems to mainly respond with `429 Too Many Requests` errors since then.

### PR DESCRIPTION
**Pull request by Scriptim**
**Description**
- A request to the Reddit API sends a user agent string that conforms to the [Reddit API rules](https://github.com/reddit-archive/reddit/wiki/API#rules).
- Posts from multiple subreddits to be sent to the same Discord text channel are all retrieved with a single request, thereby reducing the total number of requests.

**Issue**
This potentially resolves #124.

**Additional Notes**
The number of requests can be further optimized in such a way that _all_ posts are retrieved at once. In this case, the mapping to the correct Discord text channels would have to be done subsequently.

Regarding the settings for the cog, the following changes:
- A Reddit user must be set to be named as the author of the application (i. e. the bot) in the user agent.
- The limit changes its meaning from a per-subreddit limit to a per-textchannel limit.

### CODE REVIEW

File: integrations/reddit/cog.py [hunk 1]
```diff
@@ -39,16 +54,18 @@ async def get_subreddit_name(subreddit: str) -> str:
...
-async def fetch_reddit_posts(subreddit: str, limit: int) -> Optional[List[dict]]:
+async def fetch_reddit_posts(subreddits: Union[str, List[str]], limit: int) -> Optional[List[dict]]:
+    subreddits_str = subreddits if isinstance(subreddits, str) else "+".join(subreddits)
+
     async with ClientSession() as session, session.get(
         # raw_json=1 as parameter to get unicode characters instead of html escape sequences
-        f"https://www.reddit.com/r/{subreddit}/hot.json?raw_json=1",
-        headers={"User-agent": f"{Config.NAME}/{Config.VERSION}"},
+        f"https://www.reddit.com/r/{subreddits_str}/hot.json?raw_json=1",
+        headers={"User-agent": user_agent()},
```
Comment (Tert0) [inline integrations/reddit/cog.py]: `user_agent()` is an async function should it not be awaited?

File: integrations/reddit/cog.py [hunk 2]
```diff
@@ -39,16 +54,18 @@ async def get_subreddit_name(subreddit: str) -> str:
     async with ClientSession() as session, session.get(
         # raw_json=1 as parameter to get unicode characters instead of html escape sequences
         f"https://www.reddit.com/r/{subreddit}/about.json?raw_json=1",
-        headers={"User-agent": f"{Config.NAME}/{Config.VERSION}"},
+        headers={"User-agent": user_agent()},
```
Comment (Tert0) [inline integrations/reddit/cog.py]: `user_agent()` is an async function should it not be awaited?

File: integrations/reddit/cog.py [hunk 3]
```diff
@@ -26,11 +28,24 @@
...
+
+    return f"{target_platform}:{app_id}:{version_string} (by {author})"
+
+
 async def exists_subreddit(subreddit: str) -> bool:
     async with ClientSession() as session, session.get(
         # raw_json=1 as parameter to get unicode characters instead of html escape sequences
         f"https://www.reddit.com/r/{subreddit}/about.json?raw_json=1",
-        headers={"User-agent": f"{Config.NAME}/{Config.VERSION}"},
+        headers={"User-agent": user_agent()},
```
Comment (Tert0) [inline integrations/reddit/cog.py]: `user_agent()` is an async function should it not be awaited?

**Comment by Tert0** [comment]
> Note:
Don’t create a session per request. Most likely you need a session per application which performs all requests altogether.
More complex cases may require a session per site, e.g. one for Github and other one for Facebook APIs. Anyway making a session for every request is a very bad idea.
A session contains a connection pool inside. Connection reusage and keep-alives (both are on by default) may speed up total performance.

https://docs.aiohttp.org/en/stable/client_quickstart.html

File: integrations/reddit/cog.py (hunk 1 above)
Comment (Scriptim) [reply integrations/reddit/cog.py]: sure

File: integrations/reddit/cog.py [hunk 4]
```diff
@@ -39,16 +54,18 @@ async def get_subreddit_name(subreddit: str) -> str:
     async with ClientSession() as session, session.get(
```
Comment (TheCataliasTNT2k) [inline integrations/reddit/cog.py:57]: If the user is not set, no api calls should be made to reddit in the entire cog.

**Review by TheCataliasTNT2k — CHANGES_REQUESTED** [review]

File: integrations/reddit/cog.py (hunk 4 above)
Comment (Scriptim) [reply integrations/reddit/cog.py:57]: should be resolved with 17e591d

File: integrations/reddit/cog.py [hunk 5]
```diff
@@ -155,6 +161,10 @@ def text_channel_key(reddit_channel: RedditChannel) -> Optional[TextChannel]:
         await RedditPost.clean()
 
     async def start_loop(self, interval):
+        if not await is_author_set():
+            logger.warning(t.log_no_reddit_author_set)
```
Comment (Defelo) [inline integrations/reddit/cog.py]: Please send this warning into the guild's alert channel.

File: integrations/reddit/cog.py [hunk 6]
```diff
@@ -239,13 +274,28 @@ async def reddit_limit(self, ctx: Context, limit: int):
...
+    @RedditPermission.write.check
+    async def reddit_ua_author(self, ctx: Context, author: str):
+        """
+        set author (Reddit username) for user agent string
+        """
+
+        await RedditSettings.author.set(author)
+        embed = Embed(title=t.reddit, colour=Colors.Reddit, description=t.reddit_author_set)
+        await send_to_changelog(ctx.guild, t.log_reddit_author_set(author))
+        await reply(ctx, embed=embed)
```
Comment (Defelo) [inline integrations/reddit/cog.py:296]: You should start the loop here if it was previously disabled.

File: integrations/reddit/cog.py (hunk 4 above)
Comment (Defelo) [reply integrations/reddit/cog.py:57]: `exists_subreddit` and `get_subreddit_name` are still called when using `.reddit add` and `.reddit remove`, but they don't check if the user is set.

**Review by Defelo — CHANGES_REQUESTED** [review]

File: integrations/reddit/cog.py (hunk 6 above)
Comment (Scriptim) [reply integrations/reddit/cog.py:296]: Maybe we should just give a hint that the loop is not running and that one can start it with the trigger command? There might be other reasons why the loop is disabled, and in that case we shouldn't necessarily start the loop every time the author changes.

File: integrations/reddit/cog.py (hunk 5 above)
Comment (Scriptim) [reply integrations/reddit/cog.py]: 9e9f225

File: integrations/reddit/cog.py (hunk 4 above)
Comment (Scriptim) [reply integrations/reddit/cog.py:57]: abdc849

File: integrations/reddit/cog.py (hunk 6 above)
Comment (Scriptim) [reply integrations/reddit/cog.py:296]: Should be resolved with 22d4fd4

File: integrations/reddit/cog.py [hunk 7]
```diff
@@ -134,6 +164,10 @@ async def pull_hot_posts(self):
         await RedditPost.clean()
 
     async def start_loop(self, interval):
+        if not await is_author_set():
+            await send_alert(t.log_no_reddit_author_set)
```
Comment (Defelo) [inline integrations/reddit/cog.py]:
```suggestion
            await send_alert(self.bot.guilds[0], t.log_no_reddit_author_set)
```

**Review by Defelo — CHANGES_REQUESTED** [review]
> The limit changes its meaning from a per-subreddit limit to a per-textchannel limit.

Would it be possible to change that back? Or split the posts evenly between the subreddits, so that not x posts are taken from the first subreddit and the others are ignored?

**Comment by Scriptim** [comment]
> > The limit changes its meaning from a per-subreddit limit to a per-textchannel limit.
> 
> Would it be possible to change that back? Or split the posts evenly between the subreddits, so that not x posts are taken from the first subreddit and the others are ignored?

It's not that straightforward. The new API calls fetch posts from multiple subreddits at once by having the subreddits joined together in the URL with a plus. See for example [here](https://www.reddit.com/r/ProgrammerHumor+mathmemes). It is beyond our control how the posts are distributed among the different subreddits, though I guess the distribution is quite fair. In theory, you could rank the posts yourself again, so that they are evenly distributed. In this case, however, you would have to either query more posts than you actually want to send, or query posts repeatedly until you have gathered enough. In a way, this would defeat the purpose of this pull request. I don't think there is a reasonable solution to implement a per-subreddit limit without doing unnecessary API requests again. So I' d advocate leaving it at the per-textchannel limit.

File: integrations/reddit/cog.py (hunk 7 above)
Comment (Scriptim) [reply integrations/reddit/cog.py]: 09f8a50416d7b54f2c2ce48cf5430a6b519846d0

**Comment by Scriptim** [comment]
Hey, jfyi: I will not continue to work on this pull request. If you have any change requests, you'll either have to implement them yourself or simply close the pull request (if I eventually get the impression that it's going stale, I'll close it on my own).

**Pull request closed without merging**